commit 41a2a23e5da11cfbae054dcd79512fb816c37d30
Author: Jan Friesse <jfriesse@redhat.com>
Date:   Thu Feb 25 15:38:25 2010 +0100

    Cpg join with undelivered leave message
    
    Patch handles situation, when on one node, one process:
    - join cpg
    - do same actions
    - leave cpg
    - join cpg again
    
    Following sequence can (racy) end with broken process_info list.
    
    To solve this problem, one more check is done in
    message_handler_req_lib_cpg_join so if process_info with same pid and
    group as new join request exists, CPG_ERR_TRY_AGAIN is returned.

Index: corosync-1.2.0/services/cpg.c
===================================================================
--- corosync-1.2.0.orig/services/cpg.c	2010-02-26 10:20:18.000000000 +0100
+++ corosync-1.2.0/services/cpg.c	2010-02-26 10:20:32.000000000 +0100
@@ -1063,6 +1063,21 @@
 		}
 	}
 
+	/*
+	 * Same check must be done in process info list, because there may be not yet delivered
+	 * leave of client.
+	 */
+	for (iter = process_info_list_head.next; iter != &process_info_list_head; iter = iter->next) {
+		struct process_info *pi = list_entry (iter, struct process_info, list);
+
+		if (pi->nodeid == api->totem_nodeid_get () && pi->pid == req_lib_cpg_join->pid &&
+		    mar_name_compare(&req_lib_cpg_join->group_name, &pi->group) == 0) {
+			/* We have same pid and group name joined -> return error */
+			error = CPG_ERR_TRY_AGAIN;
+			goto response_send;
+		}
+	}
+
 	switch (cpd->cpd_state) {
 	case CPD_STATE_UNJOINED:
 		error = CPG_OK;
